hybrid/threads_ddc.c

```diff
@@ -0,0 +1,69 @@
+/***
+ * This example explores setting different DDCs across threads, ensuring each
+ * thread has its own DDC which does not interfere with other threads.
+ *
+ * The number of threads to execute can be changed by modifying `t_count`.
+ ***/
+
+#include <math.h>
+#include <pthread.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "cheriintrin.h"
+
+#include "../include/common.h"
+#include "./include/utils.h"
+
+struct t_info
+{
+    unsigned short count;
+    pthread_t tid;
+    void *addr_in;
+    void *__capability ddc_out;
+};
+
+void *run_thread(void *curr_t_info_ptr)
+{
+    struct t_info *curr_t_info = (struct t_info *) curr_t_info_ptr;
+    void *__capability new_ddc =
+        cheri_address_set(cheri_ddc_get(), (intptr_t) curr_t_info->addr_in);
+    write_ddc(new_ddc);
+    // Force wait to ensure threads don't die too fast
+    for (size_t i = 0; i < pow(10, 6); ++i)
+    {
+    };
+    curr_t_info->ddc_out = read_ddc();
```

Comment:
Can we add an `assert(curr_t_info->ddc_out == curr_t_info->addr_in)` or similar here?

Reply:
Done in https://github.com/capablevms/cheri-examples/pull/94/commits/66126d788c2d2ad8c04c8f499cd590f7b9a93359. Also added a check to ensure the DDC is changed.